tools/clang_format_wrapper.py

```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python
+"""Wrapper script for `git-clang-format` to avoid race conditions when
+merging/rebasing CPP files with merge conflicts."""
+from collections import namedtuple
+from os import environ as env
+from subprocess import check_output, CalledProcessError
+from typing import List
+import argparse
+import logging
+
+logging.basicConfig(format="[%(levelname)s] %(message)s", level=logging.DEBUG)
+logger = logging.getLogger(__name__)
+
+
+class GitStatus:
+    def __init__(self):
+        """GIT status output helper object.
+
+        Excerpt from https://git-scm.com/docs/git-status:
+        For paths with merge conflicts, X and Y show the modification states
+        of each side of the merge. For paths that do not have merge conflicts,
+        X shows the status of the index, and Y shows the status of the work
+        tree. For untracked paths, XY are ??. Other status codes can be
+        interpreted as follows:
+        M = modified
+        A = added
+        D = deleted
+        R = renamed
+        C = copied
+        U = updated but unmerged
+        """
+        command = ["git", "status", "--porcelain"]
+        logger.debug("Executing `%s`", " ".join(command))
+        self._git_command_output = check_output(args=command, text=True)
+
+        self.edits = self._git_command_output.rstrip().split("\n")
+        logger.debug(
+            "Found %d paths with edits in GIT status output", len(self.edits)
+        )
+        logger.debug("Pending edits:\n%s", "\n".join(self.edits))
+
+        GitEdit = namedtuple("GitEdit", ["x", "y", "path"])
+        self.edits = [
+            GitEdit(edit[0:1], edit[1:2], edit[3:]) for edit in self.edits
+        ]
+
+    @property
+    def added_and_modified_paths(self) -> List[str]:
+        return [edit.path for edit in self.edits if edit.x in ("A", "M")]
+
+
+def run_clang_format(file_list: List[str]) -> None:
+    """Execute git-clang-format for provided `file_list`."""
+    if not file_list:
+        logger.info("No file to be formatted, exiting")
+        return
+
+    command = ["git-clang-format"]
+
+    src_commit = env.get("PRE_COMMIT_FROM_REF") or env.get("PRE_COMMIT_SOURCE")
+
+    if src_commit:
+        command.append(src_commit)
+
+    command.append("--")
+    command.extend(file_list)
+
+    logger.info("Executing `%s`", " ".join(command))
+    try:
+        output = check_output(command, text=True)
+    except CalledProcessError as error:
+        logger.error(error)
+        raise SystemExit(error.returncode)
```

Comment:
don't do that here, that makes it less reusable. You could just `return error.returncode`, but IMHO we don't need to catch that error at all. If the call fails something is wrong and we want the traceback and the verbose error message (also, you didn't catch the "git status --porcelain" above, so not catching this one is more consistent).